src/js/Route.js

```diff
@@ -78,18 +80,26 @@ export default class Route {
      * @return {String}
      */
     compile(params) {
-        if (!this.parameterSegments.length) return this.template;
+        const segments = this.parameterSegments;
+        if (!segments.length) {
+            return this.template;
+        }
 
-        return this.template.replace(/{([^}?]+)\??}/g, (_, segment) => {
+        return this.template.replace(/{([^}?]+)(\??)}/g, (_, segment, optional) => {
             // If the parameter is missing but is not optional, throw an error
-            if ([null, undefined].includes(params[segment]) && this.parameterSegments.find(({ name }) => name === segment).required) {
+            if (!optional && [null, undefined].includes(params[segment])) {
                 throw new Error(`Ziggy error: '${segment}' parameter is required for route '${this.name}'.`)
             }
 
-            if (this.parameterSegments[this.parameterSegments.length - 1].name === segment && this.wheres[segment] === '.*') {
+            if (segments[segments.length - 1].name === segment && this.wheres[segment] === '.*') {
                 return encodeURIComponent(params[segment] ?? '').replaceAll('%2F', '/');
             }
 
+            if (this.wheres[segment] &&
```

Comment:
This change is just to throw an error if we are compiling the route with params that don't match the where format